sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala

```diff
@@ -62,6 +66,122 @@ private[sql] class ResolveDataSource(sparkSession: SparkSession) extends Rule[Lo
+    // config, and do various checks:
+    //   * column names in table definition can't be duplicated.
+    //   * partition, bucket and sort column names must exist in table definition.
+    //   * partition, bucket and sort column names can't be duplicated.
+    //   * can't use all table columns as partition columns.
+    //   * partition columns' type must be AtomicType.
+    //   * sort columns' type must be orderable.
+    case c @ CreateTable(tableDesc, mode, query) if c.childrenResolved =>
+      val schema = if (query.isDefined) query.get.schema else tableDesc.schema
+      checkDuplication(schema.map(_.name), "table definition of " + tableDesc.identifier)
+
+      val partitionColsChecked = checkPartitionColumns(schema, tableDesc)
+      val bucketColsChecked = checkBucketColumns(schema, partitionColsChecked)
+      c.copy(tableDesc = bucketColsChecked)
+  }
+
+  private def checkPartitionColumns(schema: StructType, tableDesc: CatalogTable): CatalogTable = {
+    val normalizedPartitionCols = tableDesc.partitionColumnNames.map { colName =>
+      normalizeColumnName(tableDesc.identifier, schema, colName, "partition")
+    }
+    checkDuplication(normalizedPartitionCols, "partition")
+
+    if (schema.nonEmpty && normalizedPartitionCols.length == schema.length) {
+      failAnalysis("Cannot use all columns for partition columns")
+    }
+
+    schema.filter(f => normalizedPartitionCols.contains(f.name)).map(_.dataType).foreach {
+      case _: AtomicType => // OK
+      case other => failAnalysis(s"Cannot use ${other.simpleString} for partition column")
+    }
+
+    tableDesc.copy(partitionColumnNames = normalizedPartitionCols)
+  }
+
+  private def checkBucketColumns(schema: StructType, tableDesc: CatalogTable): CatalogTable = {
+    tableDesc.bucketSpec match {
+      case Some(BucketSpec(numBuckets, bucketColumnNames, sortColumnNames)) =>
+        val normalizedBucketCols = bucketColumnNames.map { colName =>
+          normalizeColumnName(tableDesc.identifier, schema, colName, "bucket")
+        }
+        checkDuplication(normalizedBucketCols, "bucket")
+
+        val normalizedSortCols = sortColumnNames.map { colName =>
+          normalizeColumnName(tableDesc.identifier, schema, colName, "sort")
+        }
+        checkDuplication(normalizedSortCols, "sort")
+
+        schema.filter(f => normalizedSortCols.contains(f.name)).map(_.dataType).foreach {
+          case dt if RowOrdering.isOrderable(dt) => // OK
+          case other => failAnalysis(s"Cannot use ${other.simpleString} for sorting column")
+        }
+
+        tableDesc.copy(
+          bucketSpec = Some(BucketSpec(numBuckets, normalizedBucketCols, normalizedSortCols))
+        )
+
+      case None => tableDesc
+    }
+  }
+
+  private def checkDuplication(colNames: Seq[String], colType: String): Unit = {
+    if (colNames.distinct.length != colNames.length) {
+      val duplicateColumns = colNames.groupBy(identity).collect {
```

Comment:
Instead of using `identity`, we can do something like:
```Scala
  private def convertToCaseSensitiveAnalysisAware(name: String): String = {
    if (conf.caseSensitiveAnalysis) name else name.toLowerCase
  }

```Scala
    val duplicateColumns =
      columnNames.groupBy(convertToCaseSensitiveAnalysisAware).collect {
        case (x, ys) if ys.length > 1 => s"`$x`"
      }
```

This is from this PR: https://github.com/apache/spark/pull/13756

Do you think it is OK?